Source/CesiumRuntime/Private/CesiumGoogleMapTilesRasterOverlay.cpp

```diff
@@ -0,0 +1,124 @@
+  case EGoogleMapTilesMapType::Terrain:
+    return GoogleMapTilesMapType::terrain;
+  case EGoogleMapTilesMapType::Satellite:
+  default:
+    return GoogleMapTilesMapType::satellite;
+  }
+}
+
+std::string getScale(EGoogleMapTilesScale scale) {
+  switch (scale) {
+  case EGoogleMapTilesScale::ScaleFactor4x:
+    return GoogleMapTilesScale::scaleFactor4x;
+  case EGoogleMapTilesScale::ScaleFactor2x:
+    return GoogleMapTilesScale::scaleFactor2x;
+  case EGoogleMapTilesScale::ScaleFactor1x:
+  default:
+    return GoogleMapTilesScale::scaleFactor1x;
+  }
+}
+
+std::optional<std::vector<std::string>> getLayerTypes(
+    const TArray<EGoogleMapTilesLayerType>& layerTypes,
+    EGoogleMapTilesMapType mapType) {
+  std::vector<std::string> result;
+
+  bool hasRoadmap = false;
+
+  if (!layerTypes.IsEmpty()) {
+    result.reserve(layerTypes.Num());
+
+    for (EGoogleMapTilesLayerType layerType : layerTypes) {
+      switch (layerType) {
+      case EGoogleMapTilesLayerType::Roadmap:
+        hasRoadmap = true;
+        result.emplace_back(GoogleMapTilesLayerType::layerRoadmap);
+        break;
+      case EGoogleMapTilesLayerType::Streetview:
+        result.emplace_back(GoogleMapTilesLayerType::layerStreetview);
+        break;
+      case EGoogleMapTilesLayerType::Traffic:
+        result.emplace_back(GoogleMapTilesLayerType::layerTraffic);
+        break;
+      }
+    }
+  }
+
+  if (mapType == EGoogleMapTilesMapType::Terrain && !hasRoadmap) {
+    UE_LOG(
+        LogCesium,
+        Warning,
+        TEXT(
+            "When the MapType is set to Terrain, the LayerType must contain Roadmap."));
+  }
+
+  return result;
+}
+
+JsonValue::Array getStyles(const TArray<FString>& styles) {
+  JsonValue::Array result;
+  result.reserve(styles.Num());
+
+  JsonObjectJsonHandler handler{};
+
+  for (int32 i = 0; i < styles.Num(); ++i) {
+    const FString& style = styles[i];
+    std::string styleUtf8 = TCHAR_TO_UTF8(*style);
+    ReadJsonResult<JsonValue> response = JsonReader::readJson(
+        std::span<const std::byte>(
+            reinterpret_cast<const std::byte*>(styleUtf8.data()),
+            styleUtf8.size()),
+        handler);
+
+    ErrorList errorList;
+    errorList.errors = std::move(response.errors);
+    errorList.warnings = std::move(response.warnings);
+    errorList.log(
+        spdlog::default_logger(),
+        fmt::format("Problems parsing JSON in element {} of Styles:", i));
+
+    if (response.value) {
+      result.emplace_back(std::move(*response.value));
+    }
+  }
+
+  return result;
+}
+
+} // namespace
+
+std::unique_ptr<CesiumRasterOverlays::RasterOverlay>
+UCesiumGoogleMapTilesRasterOverlay::CreateOverlay(
+    const CesiumRasterOverlays::RasterOverlayOptions& options) {
+  return std::make_unique<CesiumRasterOverlays::GoogleMapTilesRasterOverlay>(
+      TCHAR_TO_UTF8(*this->MaterialLayerKey),
+      CesiumRasterOverlays::GoogleMapTilesNewSessionParameters{
+          .key = TCHAR_TO_UTF8(*this->Key),
```

Comment:
Can we avoid creating the raster overlay if the key is empty? Similar to how an ion raster overlay is not created if the asset ID is 0.